Gems/ROS2/Code/Source/RobotImporter/Utils/RobotImporterUtils.cpp

```diff
@@ -11,13 +11,84 @@
 #include <AzCore/Asset/AssetManager.h>
 #include <AzCore/Asset/AssetManagerBus.h>
 #include <AzCore/IO/Path/Path.h>
+#include <AzCore/Settings/SettingsRegistry.h>
 #include <AzCore/StringFunc/StringFunc.h>
 #include <AzCore/std/string/regex.h>
 #include <AzToolsFramework/API/EditorAssetSystemAPI.h>
 
 namespace ROS2
 {
 
+    bool Utils::WaitForAssetsToProcess(const AZStd::unordered_map<AZStd::string, AZ::IO::Path>& sourceAssetsPaths)
+    {
+        bool allAssetProcessed = false;
+        bool assetProcessorFailed = false;
+        auto loopStartTime = AZStd::chrono::system_clock::now();
+
+        AZStd::chrono::seconds assetJobTimeout = AZStd::chrono::seconds(30);
+        auto settingsRegistry = AZ::SettingsRegistry::Get();
+        AZ::s64 loopTimeoutValue;
+        if (settingsRegistry->Get(loopTimeoutValue, "/O3DE/ROS2/RobotImporter/AssetProcessorTimeoutInSeconds"))
+        {
+            assetJobTimeout = AZStd::chrono::seconds(loopTimeoutValue);
+        }
+
+        /* This loop waits until all of the assets are processed.
+           There are three stop conditions: allAssetProcessed, assetProcessorFailed and a timeout.
+           After all asset are processed the allAssetProcessed will be set to true.
+           assetProcessorFailed will be set to true if the asset processor does not respond.
+           The time out will break the loop if assetLoopTimeout is exceed. */
+        while (!allAssetProcessed && !assetProcessorFailed)
+        {
+            auto loopTime = AZStd::chrono::system_clock::now();
+
+            if (loopTime - loopStartTime > assetJobTimeout)
+            {
+                AZ_Warning("RobotImporterUtils", false, "Loop waiting for assets timed out");
+                break;
+            }
+
+            allAssetProcessed = true;
+            for (const auto& [AssetFileName, AssetFilePath] : sourceAssetsPaths)
+            {
+                if (AssetFilePath.empty())
+                {
+                    AZ_Warning("RobotImporterUtils", false, "WaitForAssetsToProcess got an empty filepath ");
+                    continue;
+                }
+                using namespace AzToolsFramework;
+                using namespace AzToolsFramework::AssetSystem;
+                AZ::Outcome<AssetSystem::JobInfoContainer> result = AZ::Failure();
+                AssetSystemJobRequestBus::BroadcastResult(
+                    result, &AssetSystemJobRequestBus::Events::GetAssetJobsInfo, AssetFilePath.String(), true);
```

Comment:
The Asset Processor actually has a [CompileAssetsSync](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/Asset/AssetSystemBus.h#L191-L199) API that can force any known assets based on the relative path to process synchronously

Reply:
I'm not sure if this can be applied here as we need to wait for assets that have not been detected yet by the Asset Processor.

Reply:
@lumberyard-employee-dm  is right, we should avoid blocking calls here. Since this code is working fix, I created GHI https://github.com/o3de/o3de-extras/issues/420 to test approach with `[CompileAssetsSync](https://github.com/o3de/o3de/blob/development/Code/Framework/AzFramework/AzFramework/Asset/AssetSystemBus.h#L191-L199)`